### DISCUSSION TITLE
Is root level conftest.py still recommended?

### DISCUSSION THREAD

**Discussion by pllim** [pytest-dev/pytest#9903] [Q&A]
I am not sure how but I am in this situation where pytest is going through all my source checkouts and picking up all their root-level conftest.py files and then errors out. Should I move the conftest.py file one-level down in the code tree and be done with it? I vaguely remember you recommend a root-level conftest.py but that was a while ago. Or is there another solution to ask pytest not to do this?

```
my_source_checkouts/
    package_1/
        conftest.py
        package_1/
            (codes)
    package_2/
        conftest.py
        package_2/
            (codes)
```

```
$ cd my_source_checkouts/package_2
$ pytest package_2/blahblah/tests/test_file.py
ImportError while loading conftest 'C:\...\package_2\conftest.py'.
_pytest.pathlib.ImportPathMismatchError: ('conftest', 'c:\\...\\package_1\\conftest.py', WindowsPath('C:/.../package_2/conftest.py'))
```

I am using Git Bash on Windows 11, hence the weird path formatting in the message. Same problem with pytest 7.1.2 and 7.0.x.

**Comment by portyu9** [comment]
Hi @pllim 

Yes. A root-level `conftest.py` is still a normal and supported place for project-wide fixtures/hooks.

The problem here is not that root `conftest.py` is obsolete. It is that both sibling source checkouts are visible on Python's import path, and under the traditional `prepend` import mode their `conftest.py` files can collide under the same top-level module name.

Two things I would change:

1. Make sure only the checkout you are actually testing is on `PYTHONPATH` / `sys.path`.
2. On modern pytest, prefer:

```ini
[pytest]
addopts = --import-mode=importlib
```

`importlib` mode avoids pytest modifying `sys.path` in the way `prepend` does and avoids many test-module/conftest basename collisions.

Adding `__init__.py` files can also give modules distinct package names under the older import modes, but it does not solve an environment that deliberately exposes multiple source trees under conflicting import names.

Also, `rootdir` does **not** isolate Python's import path, so changing `rootdir` alone is not the right fix.

Please mark this answer as accepted if it helped, thank you.
